platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/AdditionalDataBuilderFactory.java

```diff
@@ -16,35 +16,94 @@
+                    return DefaultDeprecationData.builder((DeprecationData) instance);
+                }
+            }))
+        .put(TypeValidationDataSpec.class, new DataTypeAndProvider(
+            TypeValidationData.class,
+            new Function<AdditionalData, AdditionalDataBuilder<? extends AdditionalData>>() {
+                @Override
+                public AdditionalDataBuilder<? extends AdditionalData> apply(@Nullable AdditionalData instance) {
+                    return DefaultTypeValidationData.builder((TypeValidationData) instance);
+                }
+            }))
+        .put(PropertyTraceDataSpec.class, new DataTypeAndProvider(
+            PropertyTraceData.class,
+            new Function<AdditionalData, AdditionalDataBuilder<? extends AdditionalData>>() {
+                @Override
+                public AdditionalDataBuilder<? extends AdditionalData> apply(@Nullable AdditionalData instance) {
+                    return DefaultPropertyTraceData.builder((PropertyTraceData) instance);
+                }
+            }))
+        .build();
+
     @SuppressWarnings("unchecked")
-    public static <S extends AdditionalData> AdditionalDataBuilder<S>  builderFor(S instance) {
-        Preconditions.checkNotNull(instance);
-        if (TypeValidationData.class.isInstance(instance)) {
-            return (AdditionalDataBuilder<S>) DefaultTypeValidationData.builder((TypeValidationData) instance);
-        } else if (DeprecationData.class.isInstance(instance)) {
-            return (AdditionalDataBuilder<S>) DefaultDeprecationData.builder((DeprecationData) instance);
-        } else if (GeneralData.class.isInstance(instance)) {
-            return (AdditionalDataBuilder<S>) DefaultGeneralData.builder((GeneralData) instance);
-        } else {
-            throw new IllegalArgumentException("Unsupported instance: " + instance);
+    private static <S extends AdditionalData, U extends AdditionalDataSpec> AdditionalDataBuilder<S> builderFor(Class<? extends U> specType, @Nullable S instance, String illegalArgumentMessage) {
+        Preconditions.checkNotNull(specType);
+        DataTypeAndProvider dataTypeAndProvider = ADDITIONAL_DATA_BUILDER_PROVIDERS.get(specType);
+        if (dataTypeAndProvider != null) {
+            return (AdditionalDataBuilder<S>) dataTypeAndProvider.builderProvider.apply(instance);
         }
+        throw new IllegalArgumentException(illegalArgumentMessage);
+    }
+
+    static <U extends AdditionalDataSpec> AdditionalDataBuilder<? extends AdditionalData> createAdditionalDataBuilder(Class<? extends U> specType, @Nullable AdditionalData additionalData) {
+        if (additionalData == null) {
+            return builderFor(specType, null, "Unsupported type: " + specType);
+        }
+        if (isCompatible(specType, additionalData)) {
+            return builderFor(specType, additionalData, "Unsupported instance: " + additionalData);
+        }
```

Comment:
🤔 I'm interested if you think this added some benefits. I always find it weird when you need to assemble an error outside the method you are doing. If you think the previous solution was more fitting, I'm okay to roll it back.

Reply:
deduplicated code.
the string was the only difference.